contracts/nftauction.js

```diff
@@ -0,0 +1,576 @@
+  if (result.errors === undefined
+    && result.events && result.events.find(el => el.contract === 'tokens'
+    && el.event === action
+    && el.data.from === from
+    && el.data.to === to
+    && api.BigNumber(el.data.quantity).eq(quantity)
+    && el.data.symbol === symbol) !== undefined) {
+    return true;
+  }
+
+  return false;
+};
+
+const settleAuction = async (auction, id = null, settle = true) => {
+  const {
+    auctionId,
+    symbol,
+    seller,
+    nftIds,
+    priceSymbol,
+    bids,
+    currentLead,
+  } = auction;
+
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const timestamp = blockDate.getTime();
+
+  const wrappedNfts = {
+    symbol,
+    ids: nftIds,
+  };
+
+  if (bids.length > 0) {
+    const bidId = id === null ? currentLead : id;
+    const leadBid = bids[bidId];
+    const finalTo = settle ? leadBid.account : seller;
+
+    // only send payment & NFTs if auction is settling (not cancelled)
+    if (settle) {
+      bids.splice(bidId, 1);
+
+      // send payment to the seller
+      await api.transferTokens(seller, priceSymbol, leadBid.bid, 'user');
+    }
+
+    // transfer the NFTs to the buyer/seller
+    await api.executeSmartContract('nft', 'transfer', {
+      fromType: 'contract',
+      to: finalTo,
+      toType: 'user',
+      nfts: [wrappedNfts],
+      isSignedWithActiveKey: true,
+    });
+
+    let count = 0;
+
+    while (count < bids.length) {
+      const {
+        account,
+        bid,
+      } = bids[count];
+
+      // send tokens back to bidders
+      await api.transferTokens(account, priceSymbol, bid, 'user');
+      count += 1;
+    }
+
+    await api.db.remove('auctions', auction);
+
+    if (settle) {
+      api.emit('settleAuction', {
+        auctionId,
+        symbol,
+        seller,
+        nftIds,
+        bidder: leadBid.account,
+        price: leadBid.bid,
+        priceSymbol,
+        timestamp,
+      });
+    } else {
+      api.emit('cancelAuction', {
+        auctionId,
+      });
+    }
+  } else {
+    // if there are no bids, expire the auction
+    await api.db.remove('auctions', auction);
+
+    // return the NFTs to the seller
+    await api.executeSmartContract('nft', 'transfer', {
+      fromType: 'contract',
+      to: seller,
+      toType: 'user',
+      nfts: [wrappedNfts],
+      isSignedWithActiveKey: true,
+    });
+
+    api.emit('expireAuction', {
+      auctionId,
+      symbol,
+      seller,
+      nftIds,
+      timestamp,
+    });
+  }
+};
+
+actions.create = async (payload) => {
+  const {
+    symbol,
+    nfts,
+    minBid,
+    finalPrice,
+    priceSymbol,
+    expiry,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(symbol && typeof symbol === 'string', 'invalid symbol')) return;
+
+  const nft = await api.db.findOneInTable('nft', 'nfts', { symbol });
+
+  if (api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')
```

Comment:
I think cleaner to just keep using !api.assert(...) return;  as above  so you don't have to nest the following code